File: backend/app/schemas/xp.py
"""
Pydantic schemas for the XP & Level system.
"""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, field_validator

from app.models.xp import XPReason


class XPTransactionResponse(BaseModel):
    id: int
    student_id: int
    amount: int
    reason: XPReason
    reference_id: Optional[str] = None
    description: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class LevelInfo(BaseModel):
    level: int
    xp_required: int
    title: str

    model_config = {"from_attributes": True}


class StudentXPDetails(BaseModel):
    """Full XP snapshot used by the student dashboard."""
    student_id: int
    total_xp: int
    current_level: int
    current_level_title: str
    # XP earned within the current level (total_xp - current.xp_required)
    xp_into_level: int
    # XP needed to reach the next level (next.xp_required - current.xp_required).
    # None when the student has hit the level cap.
    xp_for_next_level: Optional[int]
    # Cumulative XP threshold for the next level. None if at cap.
    next_level_xp_required: Optional[int]
    next_level: Optional[int]
    next_level_title: Optional[str]
    # Cosmetic theme id the student has selected. Null → default theme.
    selected_theme: Optional[str] = None
    recent_transactions: List[XPTransactionResponse]


class ThemeInfo(BaseModel):
    """One row in the theme picker. Frontend resolves theme_id → palette."""
    theme_id: str
    unlock_level: int
    unlocked: bool
    selected: bool


class ThemeListResponse(BaseModel):
    selected_theme: Optional[str] = None
    themes: List[ThemeInfo]


class ThemeSelectRequest(BaseModel):
    # Empty/null clears the selection (= use default theme). Non-empty must
    # match an unlocked theme — validated server-side.
    theme_id: Optional[str] = None


class LeaderboardEntry(BaseModel):
    """Public-safe leaderboard row — no transaction details."""
    student_id: int
    username: str
    profile_pic_url: Optional[str] = None
    grade: int
    total_xp: int
    current_level: int
    rank: int


class LeaderboardResponse(BaseModel):
    scope: str  # "class" | "grade" | "school"
    entries: List[LeaderboardEntry]


class XPAdjustmentRequest(BaseModel):
    """Admin-only manual award/deduction."""
    student_id: int
    amount: int  # may be negative
    reason: str  # mandatory free-text reason — stored in description

    @field_validator("amount")
    @classmethod
    def validate_amount(cls, v: int) -> int:
        if v == 0:
            raise ValueError("Amount must be non-zero.")
        return v

    @field_validator("reason")
    @classmethod
    def validate_reason(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("Reason is required.")
        if len(v) > 300:
            raise ValueError("Reason must be ≤ 300 characters.")
        return v

File: frontend/lib/core/models/xp.dart
// Models for the XP & Level system.
//
// Mirrors backend schemas in backend/app/schemas/xp.py. Field names follow
// the Dart camelCase convention; JSON keys stay snake_case.

class XPTransactionModel {
  final int id;
  final int studentId;
  final int amount;
  final String reason;        // ATTENDANCE, HOMEWORK_ON_TIME, ...
  final String? referenceId;
  final String? description;
  final DateTime createdAt;

  const XPTransactionModel({
    required this.id,
    required this.studentId,
    required this.amount,
    required this.reason,
    this.referenceId,
    this.description,
    required this.createdAt,
  });

  factory XPTransactionModel.fromJson(Map<String, dynamic> json) =>
      XPTransactionModel(
        id: json['id'] as int,
        studentId: json['student_id'] as int,
        amount: json['amount'] as int,
        reason: json['reason'] as String,
        referenceId: json['reference_id'] as String?,
        description: json['description'] as String?,
        createdAt: DateTime.parse(json['created_at'] as String),
      );
}

class StudentXPDetails {
  final int studentId;
  final int totalXp;
  final int currentLevel;
  final String currentLevelTitle;
  final int xpIntoLevel;
  final int? xpForNextLevel;
  final int? nextLevelXpRequired;
  final int? nextLevel;
  final String? nextLevelTitle;
  // Student's chosen cosmetic theme. Null → app default theme.
  final String? selectedTheme;
  final List<XPTransactionModel> recentTransactions;

  const StudentXPDetails({
    required this.studentId,
    required this.totalXp,
    required this.currentLevel,
    required this.currentLevelTitle,
    required this.xpIntoLevel,
    this.xpForNextLevel,
    this.nextLevelXpRequired,
    this.nextLevel,
    this.nextLevelTitle,
    this.selectedTheme,
    required this.recentTransactions,
  });

  /// Fraction in [0,1] of the way through the current level. Returns 1.0
  /// at the level cap (no next level).
  double get progressToNextLevel {
    final needed = xpForNextLevel;
    if (needed == null || needed <= 0) return 1.0;
    return (xpIntoLevel / needed).clamp(0.0, 1.0);
  }

  factory StudentXPDetails.fromJson(Map<String, dynamic> json) =>
      StudentXPDetails(
        studentId: json['student_id'] as int,
        totalXp: json['total_xp'] as int,
        currentLevel: json['current_level'] as int,
        currentLevelTitle: json['current_level_title'] as String,
        xpIntoLevel: json['xp_into_level'] as int,
        xpForNextLevel: json['xp_for_next_level'] as int?,
        nextLevelXpRequired: json['next_level_xp_required'] as int?,
        nextLevel: json['next_level'] as int?,
        nextLevelTitle: json['next_level_title'] as String?,
        selectedTheme: json['selected_theme'] as String?,
        recentTransactions: ((json['recent_transactions'] as List?) ?? const [])
            .map((e) => XPTransactionModel.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

/// Catalogue entry for a single cosmetic theme. Frontend resolves
/// `themeId` → `BrandPalette` via `BrandPalettes.byId`.
class ThemeInfo {
  final String themeId;
  final int unlockLevel;
  final bool unlocked;
  final bool selected;

  const ThemeInfo({
    required this.themeId,
    required this.unlockLevel,
    required this.unlocked,
    required this.selected,
  });

  factory ThemeInfo.fromJson(Map<String, dynamic> json) => ThemeInfo(
        themeId: json['theme_id'] as String,
        unlockLevel: json['unlock_level'] as int,
        unlocked: json['unlocked'] as bool,
        selected: json['selected'] as bool,
      );
}

class ThemeListResponse {
  final String? selectedTheme;
  final List<ThemeInfo> themes;

  const ThemeListResponse({this.selectedTheme, required this.themes});

  factory ThemeListResponse.fromJson(Map<String, dynamic> json) =>
      ThemeListResponse(
        selectedTheme: json['selected_theme'] as String?,
        themes: ((json['themes'] as List?) ?? const [])
            .map((e) => ThemeInfo.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class LeaderboardEntry {
  final int rank;
  final int studentId;
  final String username;
  final String? profilePicUrl;
  final int grade;
  final int totalXp;
  final int currentLevel;

  const LeaderboardEntry({
    required this.rank,
    required this.studentId,
    required this.username,
    this.profilePicUrl,
    required this.grade,
    required this.totalXp,
    required this.currentLevel,
  });

  factory LeaderboardEntry.fromJson(Map<String, dynamic> json) =>
      LeaderboardEntry(
        rank: json['rank'] as int,
        studentId: json['student_id'] as int,
        username: json['username'] as String,
        profilePicUrl: json['profile_pic_url'] as String?,
        grade: json['grade'] as int,
        totalXp: json['total_xp'] as int,
        currentLevel: json['current_level'] as int,
      );
}

class LeaderboardResponse {
  final String scope;       // 'class' | 'grade' | 'school'
  final List<LeaderboardEntry> entries;

  const LeaderboardResponse({required this.scope, required this.entries});

  factory LeaderboardResponse.fromJson(Map<String, dynamic> json) =>
      LeaderboardResponse(
        scope: json['scope'] as String,
        entries: ((json['entries'] as List?) ?? const [])
            .map((e) => LeaderboardEntry.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class LevelInfo {
  final int level;
  final int xpRequired;
  final String title;

  const LevelInfo({
    required this.level,
    required this.xpRequired,
    required this.title,
  });

  factory LevelInfo.fromJson(Map<String, dynamic> json) => LevelInfo(
        level: json['level'] as int,
        xpRequired: json['xp_required'] as int,
        title: json['title'] as String,
      );
}
